### DISCUSSION TITLE
Why an async Server Component as child of a Client Component fails on Next.js 13 app directory?

### DISCUSSION THREAD

**Discussion by clacladev** [vercel/next.js#46931] [Help]
### Summary

I am experimenting with Next.js 13 app directory and React 18's Server Components. Following the documentation I put an async fetch() call inside a Server Component and I mark the component as async.

But then when I place it as a child of a Client Component, it fails with error `Objects are not valid as a React child (found: [object Promise])`.

If I place it as a child of another Server Component, it works as intended.

Why?

<img width="890" alt="Screenshot 2023-03-08 at 16 30 04" src="https://user-images.githubusercontent.com/161903/223774551-d4acf6ab-8f03-4222-ae57-e17689f34009.png">

Code here:
```ts
Code here:

### page.tsx
import { ClientComp } from './ClientComp'

export default function Page() {
  return <ClientComp />
}

### ClientComp.tsx
"use client";
import { ServerComp } from "./ServerComp";

export function ClientComp() {
  {/* @ts-expect-error Async Server Component */}
  return <ServerComp />
}

### ServerComp.tsx
export async function ServerComp() {
  return <h2>ServerComp</h2>
}
```

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 19.7.0
      npm: 9.5.0
      Yarn: 1.22.19
      pnpm: 7.29.1
    Relevant packages:
      next: 13.2.3
      eslint-config-next: 13.2.3
      react: 18.2.0
      react-dom: 18.2.0
```

**Comment by clacladev** [comment]
The idea is to be able to pass information/state/callbacks from the Client component to the child Server component. So adding `{children}` in the JSX of ClientComponent, would not cut it.

**Comment by icyJoseph** [comment]
Right, this is one the first things I went to try, and since then I've gotten to think that the problem is that:

- you can't pass props from client to server component

Because this is essentially a request that has to leave the client, with serialized data, and be used on the server. 

Next.js implemented half of this feature, at page level, with GetServerSideProps, which is a server side function that returns data as props. 

However, we'd still need to serialize the props, and fly the request to the server, which in turn should respond with a diff object (?), to be used by the client side renderer to adjust the UI. Which is not what GSSP does during SPA transitions.

So it is not really impossible, but it is not as easy as just passing props... perhaps a library can implement this?

A server component's HTML is sort of set in stone, that is without interactivity. 

- you can't pass callbacks to the server component

Because, this would need to leave the client, and functions/closures and other data are essentially destroyed during serialization.

## Must we really leave the client to pass props? 

Because a server component access a whole different set of API's and constraints regarding data private to the module, where the server component lives, we cannot just ship "client-like-code-for-a-server-component", without going down a path filled with pitfalls, and no clear way out.

**Reply by clacladev** [reply]
So the whole idea of trying to make all the api calls inside server components is not realistic, right?

I mean if I create a client component, eg a paginated table, then when the users clicks to see the next page, the only option I have is to make the api call from the client component. 
I technically cannot let a server component do it

**Reply by icyJoseph** [reply]
> So the whole idea of trying to make all the api calls inside server components is not realistic, right?

If they depend on client-side inputs, perhaps not, but otherwise, it is the go to tool.

> I mean if I create a client component, eg a paginated table, then when the users clicks to see the next page, the only option I have is to make the api call from the client component.

You'd need to use the URL I think. Like when the user clicks on page 3, you'd navigate to `/the-path/?page=3`, which makes your URL shareable too.

**Reply by clacladev** [reply]
Yes that what I thought. 
This makes server components async functionality much less useful in real world app I think. The majority of api calls rely on some user input or state

**Reply by icyJoseph** [reply]
Yeah, well, unless we are missing some sleight of hand, master trick, or some obvious oversight, that's what it looks like now... I think when the mutation API comes around, we'll have to revisit this.

**Reply by clacladev** [reply]
Thanks for confirming all this to me 🏋️
